tools/state-viewer/src/state_parts.rs

```diff
@@ -0,0 +1,432 @@
+use crate::epoch_info::iterate_and_filter;
+use clap::Subcommand;
+use near_chain::{ChainStore, ChainStoreAccess, RuntimeAdapter};
+use near_epoch_manager::EpochManager;
+use near_primitives::epoch_manager::epoch_info::EpochInfo;
+use near_primitives::state_part::PartId;
+use near_primitives::syncing::get_num_state_parts;
+use near_primitives::types::EpochId;
+use near_primitives_core::hash::CryptoHash;
+use near_primitives_core::types::{BlockHeight, EpochHeight, ShardId};
+use near_store::Store;
+use nearcore::{NearConfig, NightshadeRuntime};
+use s3::serde_types::ListBucketResult;
+use std::fs::DirEntry;
+use std::path::{Path, PathBuf};
+use std::str::FromStr;
+use std::sync::Arc;
+use std::time::Instant;
+
+#[derive(Subcommand, Debug, Clone)]
+pub(crate) enum EpochSelection {
+    /// Current epoch.
+    Current,
+    /// Fetch the given epoch.
+    EpochId { epoch_id: String },
+    /// Fetch epochs at the given height.
+    EpochHeight { epoch_height: EpochHeight },
+    /// Fetch an epoch containing the given block hash.
+    BlockHash { block_hash: String },
+    /// Fetch an epoch containing the given block height.
+    BlockHeight { block_height: BlockHeight },
+}
+
+impl EpochSelection {
+    pub fn to_epoch_id(
+        &self,
+        store: Store,
+        chain_store: &ChainStore,
+        epoch_manager: &EpochManager,
+    ) -> EpochId {
+        match self {
+            EpochSelection::Current => {
+                epoch_manager.get_epoch_id(&chain_store.head().unwrap().last_block_hash).unwrap()
+            }
+            EpochSelection::EpochId { epoch_id } => {
+                EpochId(CryptoHash::from_str(&epoch_id).unwrap())
+            }
+            EpochSelection::EpochHeight { epoch_height } => {
+                // Fetch epochs at the given height.
+                // There should only be one epoch at a given height. But this is a debug tool, let's check
+                // if there are multiple epochs at a given height.
+                let epoch_ids = iterate_and_filter(store, |epoch_info| {
+                    epoch_info.epoch_height() == *epoch_height
+                });
+                assert_eq!(epoch_ids.len(), 1, "{:#?}", epoch_ids);
+                epoch_ids[0].clone()
+            }
+            EpochSelection::BlockHash { block_hash } => {
+                let block_hash = CryptoHash::from_str(&block_hash).unwrap();
+                epoch_manager.get_epoch_id(&block_hash).unwrap()
+            }
+            EpochSelection::BlockHeight { block_height } => {
+                // Fetch an epoch containing the given block height.
+                let block_hash = chain_store.get_block_hash_by_height(*block_height).unwrap();
+                epoch_manager.get_epoch_id(&block_hash).unwrap()
+            }
+        }
+    }
+}
+
+/// Returns block hash of the last block of an epoch preceding the given `epoch_info`.
+fn get_prev_hash_of_epoch(
```

Comment:
We may have this info or parts of it in the storage e.g. BlockInfo looks promising.

Reply:
Not sure what you mean. I use `BlockInfo` in this loop to get `epoch_first_block()` and so on.

Reply:
Ah, nevermind then, I thought this information may be available directly (without loop) in that table.